web/src/pages/Docs.tsx

```diff
@@ -4,189 +4,168 @@
-        hideControls ? '?hide-ui=true' : ''
-      }`
-
-      // skip updating the route if the new state is the same as the current one
-      if (window.location.hash === newState.substring(1)) {
-        return
-      }
+  const updateURL = (newProject: string, newVersion: string, newPage: string, newHideUi: boolean): void => {
+    const url = `#/${newProject}/${newVersion}/${newPage}${newHideUi ? '?hide-ui=true' : ''}`
 
-      window.history.pushState({}, '', newState)
-    },
-    []
-  )
+    if (project === newProject && version === newVersion && page === newPage && hideUi === newHideUi) {
+      // no change
+      return
+    }
 
-  updateRoute(project, version, page, hideUi)
+    const oldVersion = version
+    const oldPage = page
 
-  useEffect(() => {
-    if (project === '' || project === 'none') {
-      setVersions([])
+    setProject(newProject)
+    setVersion(newVersion)
+    setPage(newPage)
+    setHideUi(newHideUi)
+
+    if (oldVersion === 'latest' && newVersion !== 'latest') {
+      // replace the url if 'latest' was updated to the actual version
+      window.history.replaceState(null, '', url)
       return
     }
 
-    if (projects == null || projects.length === 0) {
+    if (oldPage === '' && newPage !== '') {
+      // replace the url if the page was updated from '' to the actual page
+      window.history.replaceState(null, '', url)
       return
     }
 
-    try {
-      const matchingProjects = projects.filter((p) => p.name === project)
-
-      if (matchingProjects.length !== 1) {
-        setLoadingFailed(true)
-        return
-      }
-
-      let res = matchingProjects[0].versions
-
-      if (res.length === 0) {
-        setLoadingFailed(true)
-        return
-      }
-
-      res = res.sort((a, b) => ProjectRepository.compareVersions(a, b))
-      setVersions(res)
-
-      if (version !== 'latest') {
-        // custom version -> check if it exists
-        const versionsAndTags = res.map((v) => [v.name, ...v.tags]).flat()
-
-        if (!versionsAndTags.includes(version)) {
-          // version does not exist -> fail
-          setLoadingFailed(true)
-          console.log("Version doesn't exist")
-        }
-
-        return
-      }
+    window.history.pushState(null, '', url)
+  }
 
-      // latest version -> check if there is a latest tag
-      const versionWithLatestTag = res.find((v) =>
-        (v.tags ?? []).includes('latest')
-      )
-
-      // if there is a latest tag, use it,
-      // otherwise use the latest version by sorting
-      const latestVersion =
-        versionWithLatestTag != null
-          ? versionWithLatestTag.name
-          : res[res.length - 1].name
-
-      setVersion(latestVersion)
-      updateRoute(project, latestVersion, page, hideUi)
-    } catch (e) {
-      console.error(e)
-      setLoadingFailed(true)
+  const onIFrameLocationChanged = (url: string): void => {
+    url = url.split('/doc/')[1]
+    if (url.length === 0) {
+      // should never happen
```

Comment:
log an error here?

Reply:
I mean we could, but I didn't think it would be interesting